### PR TITLE
Implement Lower-Dimensional Storages for gtc:cuda Backend

### LINKED ISSUE #390: `test_lower_dimensional_inputs` fails for gtc:cuda backend
The `test_lower_dimensional_inputs` fails for gtc:cuda backend, they are marked as `pytest.xfail` for that backend. (See #388)

### PR DESCRIPTION
**Pull request by fthaler**
## Description

Support for lower-dimensional storages in the gtc:cuda backend and proper multi-loop extent analysis. Fixes #390.

## Requirements

Before submitting this PR, please make sure:

- [x] The code builds cleanly without new errors or warnings
- [x] The code passes all the existing tests
- [x] If this PR adds a new feature, new tests have been added to test these
new features
- [x] All relevant documentation has been updated or added

Additionally, if this PR contains code authored by new contributors:

- [x] All the authors are covered by a valid contributor assignment agreement,
signed by the employer if needed, provided to ETH Zurich
- [x] The names of all the new contributors have been added to an updated
version of the AUTHORS.rst file included in the PR

### CODE REVIEW

**Comment by fthaler** [comment]
Fixes #390.

File: tests/test_integration/test_code_generation.py [hunk 1]
```diff
@@ -156,9 +156,6 @@ def stencil(field_in: gtscript.Field[np.float_], field_out: gtscript.Field[np.fl
 
 @pytest.mark.parametrize("backend", ALL_BACKENDS)
 def test_lower_dimensional_inputs(backend):
-    if backend == "gtc:cuda":
-        pytest.xfail("gtc:cuda backend does not support lower dimensional fields")
-
```
Comment (havogt) [inline tests/test_integration/test_code_generation.py:161]: This test seems to be in the wrong file, no? It's a full test, not only code generation. Not your fault, though...

Comment (fthaler) [reply tests/test_integration/test_code_generation.py:161]: Yes, you are right. Should probably live in `tests/test_integration/test_suites.py`or so…

**Review by havogt — APPROVED** [review]

**Comment by havogt** [comment]
bors try

**Pull request merged**